File: docs-examples/example-java/src/test/java/io/micronaut/nats/docs/consumer/custom/annotation/SIDAnnotationBinder.java
package io.micronaut.nats.docs.consumer.custom.annotation;

import io.micronaut.context.annotation.Requires;
// tag::imports[]
import io.micronaut.core.convert.ArgumentConversionContext;
import io.micronaut.core.convert.ConversionService;
import io.micronaut.nats.bind.NatsAnnotatedArgumentBinder;
import io.nats.client.Message;
import jakarta.inject.Singleton;
// end::imports[]

@Requires(property = "spec.name", value = "SIDSpec")
// tag::clazz[]
@Singleton // <1>
public class SIDAnnotationBinder implements NatsAnnotatedArgumentBinder<SID> { // <2>

    private final ConversionService<?> conversionService;

    public SIDAnnotationBinder(ConversionService<?> conversionService) { // <3>
        this.conversionService = conversionService;
    }

    @Override
    public Class<SID> getAnnotationType() {
        return SID.class;
    }

    @Override
    public BindingResult<Object> bind(ArgumentConversionContext<Object> context, Message source) {
        String sid = source.getSID(); // <4>
        return () -> conversionService.convert(sid, context); // <5>
    }
}
// end::clazz[]

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/nats/docs/consumer/custom/annotation/SIDAnnotationBinder.kt
package io.micronaut.nats.docs.consumer.custom.annotation

import io.micronaut.context.annotation.Requires
// tag::imports[]
import io.micronaut.core.bind.ArgumentBinder
import io.micronaut.core.convert.ArgumentConversionContext
import io.micronaut.core.convert.ConversionService
import io.micronaut.nats.bind.NatsAnnotatedArgumentBinder
import io.nats.client.Message
import jakarta.inject.Singleton
// end::imports[]

@Requires(property = "spec.name", value = "SIDSpec")
// tag::clazz[]
@Singleton // <1>
class SIDAnnotationBinder(private val conversionService: ConversionService) // <3>
    : NatsAnnotatedArgumentBinder<SID> { // <2>

    override fun getAnnotationType(): Class<SID> {
        return SID::class.java
    }

    override fun bind(context: ArgumentConversionContext<Any>, source: Message): ArgumentBinder.BindingResult<Any> {
        val sid = source.sid // <4>
        return ArgumentBinder.BindingResult { conversionService.convert(sid, context) } // <5>
    }
}
// end::clazz[]
